src/api/handlers/projects/jobs.py

```diff
@@ -383,6 +383,220 @@ def get(self, project_id, job_id):
+
+        jobs = g.db.execute_many_dict('''
+            SELECT state, id, dependencies, restarted
+            FROM job
+            WHERE build_id = %s
+            AND project_id = %s
+        ''', [build_id, project_id])
+
+        restart_jobs = [job_id]
+
+        while True:
+            found = False
+            for j in jobs:
+                if j['id'] in restart_jobs:
+                    continue
+
+                if not j['dependencies']:
+                    continue
+
+                if j['restarted']:
+                    continue
+
+                for dep in j['dependencies']:
+                    dep_id = dep['job-id']
+
+                    if dep_id in restart_jobs:
+                        found = True
+                        restart_jobs.append(j['id'])
+                        break
+
+            if not found:
+                break
+
+        # get all jobs, filter the jobs marked to be restarted, check the status is running or not
+        for j in jobs:
+            if j['id'] not in restart_jobs:
+                continue
+
+            if j['state'] not in restart_states and j['state'] not in ('skipped', 'queued'):
+                abort(400, 'Some children jobs are still running')
+
+        # print mssage for the restart operator
+        if user_id is not None:
+            result = g.db.execute_one_dict('''
+                SELECT username
+                FROM "user"
+                WHERE id = %s
+            ''', [user_id])
+            username_or_token = result['username']
+        else:
+            result = g.db.execute_one_dict('''
+                                SELECT description
+                                FROM auth_token
+                                WHERE id = %s
+                            ''', [g.token['id']])
+            username_or_token = "project token " + result['description']
+        msg = 'Job restarted by %s\n' % username_or_token
+
+        # Clone Jobs and adjust dependencies
+        # get jobs that marked as restart from table job
+        jobs = []
+        for j in restart_jobs:
+            jobs += g.db.execute_many_dict('''
+                SELECT id, build_id, type, dockerfile, name, project_id, dependencies,
+                       repo, env_var, env_var_ref, build_arg, deployment, definition, cluster_name
+                FROM job
+                WHERE id = %s;
+            ''', [j])
+        logger.debug('## get jobs that marked as restart from table job:')
+        logger.debug(str(jobs))
+
+        # get job thet restart it self only.
+        single_job = []
+        single_job += g.db.execute_many_dict('''
+            SELECT id, build_id, type, dockerfile, name, project_id, dependencies,
+                    repo, env_var, env_var_ref, build_arg, deployment, definition, cluster_name
+            FROM job
+            WHERE id = %s;
+        ''', [job_id])
+
+        old_id_job = {}
+        loop_jobs = single_job
+        for j in loop_jobs:
```

Comment:
`loop_jobs` is not needed any more since we won't restart children jobs.

Reply:
thanks Berry, I will remove this useless variable